clippy_lints/src/option_if_let_else.rs

```diff
@@ -0,0 +1,176 @@
+use crate::utils::sugg::Sugg;
+use crate::utils::{match_qpath, match_type, paths, span_lint_and_sugg};
+use if_chain::if_chain;
+
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:**
+    /// Detects people who use `if let Some(v) = ... { y } else { x }`
+    /// when they could use `Option::map_or` or `Option::map_or_else`
+    /// instead.
+    ///
+    /// **Why is this bad?**
+    /// Using the dedicated function in the Option class is clearer and
+    /// more concise than an if let expression.
+    ///
+    /// **Known problems:**
+    /// Using `Option::map_or` (as the lint presently suggests) requires
+    /// it to compute the else value every function call, even if the
+    /// value is unneeded. It might be better to use Option::map_or_else
+    /// if the code is either very expensive or if it changes some state
+    /// (i.e. mutates a value). For example,
+    /// ```rust
+    /// # let optional = Some(0);
+    /// let _ = if let Some(x) = optional {
+    ///     x
+    /// } else {
+    ///     let y = 0;
+    ///     /* Some expensive computation or state-changing operation */
+    ///     y
+    /// };
+    /// ```
+    /// would be better as
+    /// ```rust
+    /// # let optional = Some(0);
+    /// let _ = optional.map_or_else(|| { let y = 0; /* some expensive computation or
+    /// state-changing operation */ y }, |x| x);
+    /// ```
+    /// instead of the suggested
+    /// ```rust
+    /// # let optional = Some(0);
+    /// let _ = optional.map_or({ let y = 0; /* some expensive computation or state-changing
+    /// operation */ y }, |x| x);
+    /// ```
```

Comment:
Instead of explaining this, you can refer to the documentation of the [`or_fun_call`](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call) lint, and note that this should be caught by that lint.

Also with your recent changes, this isn't that much of a problem anymore, since when complex blocks are involved `map_or_else` is suggested anyway.